scalikejdbc-streams/src/main/scala/scalikejdbc/streams/StreamingDB.scala

```diff
@@ -0,0 +1,288 @@
+     */
+    private[StreamingDB] var deferredError: Throwable = null
+
+    /**
+     * The state for a suspended streaming action. Must only be set from a synchronous action
+     * context.
+     */
+    private[StreamingDB] var streamState: AnyRef = null
+
+    /** The streaming action which may need to be continued with the suspended state */
+    private[StreamingDB] var streamingAction: StreamingAction[_, _ <: WithExtractor] = null
+
+    @volatile private[this] var cancelRequested = false
+
+    /** The Promise to complete when streaming has finished. */
+    val streamingResultPromise: Promise[Null] = Promise[Null]()
+
+    /**
+     * Indicate that the specified number of elements has been delivered. Returns the remaining
+     * demand. This is an atomic operation. It must only be called from the synchronous action
+     * context which performs the streaming.
+     */
+    def delivered(num: Long): Long = remaining.addAndGet(-num)
+
+    /**
+     * Get the current demand that has not yet been marked as delivered and mark it as being in
+     * the current batch. When this value is negative, the initial streaming action is still
+     * running and the real demand can be computed by subtracting `Long.MinValue` from the
+     * returned value.
+     */
+    def demandBatch: Long = remaining.get()
+
+    /** Whether the stream has been cancelled by the Subscriber */
+    def cancelled: Boolean = cancelRequested
+
+    def emit(v: Any): Unit = subscriber.asInstanceOf[Subscriber[Any]].onNext(v)
+
+    /**
+     * Finish the stream with `onComplete` if it is not finished yet. May only be called from a
+     * synchronous action context.
+     */
+    def tryOnComplete(): Unit = if (!finished && !cancelRequested) {
+      if (log.isDebugEnabled) log.debug("Signaling onComplete()")
+      finished = true
+      try subscriber.onComplete() catch {
+        case NonFatal(ex) => log.warn("Subscriber.onComplete failed unexpectedly", ex)
+      }
+    }
+
+    /**
+     * Finish the stream with `onError` if it is not finished yet. May only be called from a
+     * synchronous action context.
+     */
+    def tryOnError(t: Throwable): Unit = if (!finished) {
+      if (log.isDebugEnabled) log.debug(s"Signaling onError($t)")
+      finished = true
+      try subscriber.onError(t) catch {
+        case NonFatal(ex) => log.warn("Subscriber.onError failed unexpectedly", ex)
+      }
+    }
+
+    /** Restart a suspended streaming action. Must only be called from the Subscriber context. */
+    def restartStreaming(): Unit = {
+      val _ = sync
+      val s = streamState
+      if (s ne null) {
+        streamState = null
+        if (log.isDebugEnabled) log.debug("Scheduling stream continuation after transition from demand = 0")
+        val a = streamingAction
+        database.scheduleSynchronousStreaming(a, this.asInstanceOf[database.StreamingActionContext])(s.asInstanceOf[a.State])
```

Comment:
uh, Concurrency control is hard. So I am closing the access scope with inner-class. (it is following Slick)
now, I am trying to rewrite like:

A.
https://github.com/yoskhdia/scalikejdbc/compare/merge-scalikejdbc-streams...yoskhdia:merge-scalikejdbc-streams-inner-to-toplevel

B.
https://github.com/yoskhdia/scalikejdbc/compare/merge-scalikejdbc-streams...yoskhdia:merge-scalikejdbc-streams-separate-to-component

However, in B pattern, can not avoid asInstanceOf...